mqtt/mqtt-broker/src/broker.rs

```diff
@@ -797,6 +814,99 @@ where
 
         Ok(())
     }
+
+    async fn notify_subscription_change(&mut self, client_id: &ClientId) {
+        let topics: String = self
+            .sessions
+            .get(client_id) // find session that changed
+            .and_then(|session| match session {
+                // get session state
+                Session::Transient(connected) => Some(connected.state()),
+                Session::Persistent(connected) => Some(connected.state()),
+                Session::Offline(offline) => Some(offline.state()),
+                _ => None,
+            })
+            .map(|state| state.subscriptions().keys()) // get topic filters
+            .map(|topics| {
+                topics
+                    .map(|t| t.as_ref())
+                    .collect::<Vec<&str>>()
+                    .join(r"\u{0000}")
+            }) // join to string for payload
+            .unwrap_or_default(); // no subscriptions is default
+
+        let publication = proto::Publication {
+            topic_name: format!("$sys/subscriptions/{}", client_id),
+            qos: proto::QoS::AtMostOnce, //no ack
+            retain: true,
+            payload: Bytes::from(topics),
+        };
+
+        self.publish_all(publication).await.unwrap();
+    }
+
+    async fn notify_disconnect(&mut self, client_id: &ClientId) {
+        if let None = self.sessions.get(client_id) { // If transient session is closed, remove its subscriptions
+            let publication = proto::Publication {
+                topic_name: format!("$sys/subscriptions/{}", client_id),
+                qos: proto::QoS::AtMostOnce, //no ack
+                retain: true,
+                payload: Bytes::from("".to_owned()),
+            };
+
+            self.publish_all(publication).await.unwrap();
+        }
+
+        self.notify_connection_change().await;
+    }
+
+    async fn notify_connection_change(&mut self) {
+        let connected: String = self
+            .sessions
+            .iter()
+            .filter_map(|(client_id, session)| match session {
+                Session::Transient(_) => Some(client_id),
+                Session::Persistent(_) => Some(client_id),
+                _ => None,
+            })
+            .map(|client_id| client_id.as_str())
+            .collect::<Vec<&str>>()
+            .join(r"\u{0000}");
+
+        let publication = proto::Publication {
+            topic_name: "$sys/connected".to_owned(),
+            qos: proto::QoS::AtMostOnce, //no ack
+            retain: true,
+            payload: Bytes::from(connected),
+        };
+
+        self.publish_all(publication).await.unwrap();
```

Comment:
are you planning to implement todos (error handling specifically) in the same PR? if so, maybe move it to draft for now?

I think at least we need to write to stderr, but @dmolokanov can advice more.

Reply:
Yea, I want to do it in this pr, its a simple change I just wasn't sure the approach.